internal/printer/print-to-json.go

```diff
@@ -0,0 +1,241 @@
+package printer
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	. "github.com/withastro/compiler/internal"
+	"github.com/withastro/compiler/internal/loc"
+	"github.com/withastro/compiler/internal/sourcemap"
+	"github.com/withastro/compiler/internal/t"
+	"github.com/withastro/compiler/internal/transform"
+)
+
+type ASTPosition struct {
+	Start ASTPoint `json:"start,omitempty"`
+	End   ASTPoint `json:"end,omitempty"`
+}
+
+type ASTPoint struct {
+	Line   int `json:"line,omitempty"`
+	Column int `json:"column,omitempty"`
+	Offset int `json:"offset,omitempty"`
+}
+
+type ASTNode struct {
+	Type       string      `json:"type"`
+	Name       string      `json:"name,omitempty"`
+	Value      string      `json:"value,omitempty"`
+	Attributes []ASTNode   `json:"attributes,omitempty"`
+	Directives []ASTNode   `json:"directives,omitempty"`
+	Children   []ASTNode   `json:"children,omitempty"`
+	Position   ASTPosition `json:"position,omitempty"`
+
+	// Attributes only
+	Kind string `json:"kind,omitempty"`
+}
+
+func escapeForJSON(value string) string {
+	newlines := regexp.MustCompile(`\n`)
+	value = newlines.ReplaceAllString(value, `\n`)
+	doublequotes := regexp.MustCompile(`"`)
+	value = doublequotes.ReplaceAllString(value, `\"`)
+	amp := regexp.MustCompile(`&`)
+	value = amp.ReplaceAllString(value, `\&`)
+	r := regexp.MustCompile(`\r`)
+	value = r.ReplaceAllString(value, `\r`)
+	t := regexp.MustCompile(`\t`)
+	value = t.ReplaceAllString(value, `\t`)
+	f := regexp.MustCompile(`\f`)
+	value = f.ReplaceAllString(value, `\f`)
+	return value
+}
+
+func (n ASTNode) String() string {
+	str := fmt.Sprintf(`{"type":"%s"`, n.Type)
+	if n.Kind != "" {
+		str += fmt.Sprintf(`,"kind":"%s"`, n.Kind)
+	}
+	if n.Name != "" {
+		str += fmt.Sprintf(`,"name":"%s"`, escapeForJSON(n.Name))
+	}
+	if n.Value != "" || n.Type == "attribute" {
+		str += fmt.Sprintf(`,"value":"%s"`, escapeForJSON(n.Value))
+	}
+	if len(n.Attributes) > 0 {
+		str += `,"attributes":[`
+		for i, attr := range n.Attributes {
+			str += attr.String()
+			if i < len(n.Attributes)-1 {
+				str += ","
+			}
+		}
+		str += `]`
+	}
+	if len(n.Directives) > 0 {
+		str += `,"directives":[`
+		for i, attr := range n.Directives {
+			str += attr.String()
+			if i < len(n.Directives)-1 {
+				str += ","
+			}
+		}
+		str += `]`
+	}
+	if len(n.Children) > 0 {
+		str += `,"children":[`
+		for i, node := range n.Children {
+			str += node.String()
+			if i < len(n.Children)-1 {
+				str += ","
+			}
+		}
+		str += `]`
+	}
+	if n.Position.Start.Line != 0 {
+		str += `,"position":{`
+		str += fmt.Sprintf(`"start":{"line":%d,"column":%d,"offset":%d}`, n.Position.Start.Line, n.Position.Start.Column, n.Position.Start.Offset)
+		if n.Position.End.Line != 0 {
+			str += fmt.Sprintf(`,"end":{"line":%d,"column":%d,"offset":%d}`, n.Position.End.Line, n.Position.End.Column, n.Position.End.Offset)
+		}
+		str += "}"
+	}
+	str += "}"
+	return str
+}
```

Comment:
Just want to note that our `String()` function is implemented manually, but recurses through the AST just fine.

I had originally used `json.Marshal` but it relies heavily on Reflection which is not well supported in `tinygo`. This is more performant, anyway.